File: 백준/Gold/2239. 스도쿠/스도쿠.py
import sys
input = sys.stdin.readline

arr = [list(map(int,list(input().strip()))) for _ in range(9)]

# 우리가 채워야될 곳의 좌표를 q에 저장
q = []
for i in range(9):
    for j in range(9):
        if arr[i][j] == 0:
            q.append([i,j])

def checkrow(x,i):
    for j in range(9):
        if i == arr[x][j]:
            return False
    return True

def checkcol(y,i):
    for j in range(9):
        if i == arr[j][y]:
            return False
    return True

def checkdiag(x,y,i):
    nx = x//3*3
    ny = y//3*3
    for j in range(3):
        for k in range(3):
            if i == arr[nx+j][ny+k]:
                return False
    return True

# cnt : 빈곳을 올바른 숫자로 넣은 개수
def dfs(cnt):
    # 모든 빈칸을 올바른 숫자로 넣었다면 그래프 출력
    if cnt == len(q):
        for i in range(9):
            print(*arr[i],sep='')
        exit()    

    # 0에 1~9중 어떤 값을 넣을지 체크한다.
    for i in range(1,10):
        x = q[cnt][0]
        y = q[cnt][1]
        
        # 행, 열, 3x3 행열에 모두 조건이 맞다면
        if checkrow(x,i) and checkcol(y,i) and checkdiag(x,y,i):
            # 해당 값을 추가해준다.
            arr[x][y] = i
            dfs(cnt+1)
            # 만약 어떠한 값을 넣고 / 다음 구간들을 탐색하는데 조건에 맞는 값이 없다면?
            # 조건에 맞는 마지막 값을 0으로 바꾸고 -> 다른 값으로 다시 탐색한다.
            arr[x][y] = 0

dfs(0)

File: 백준/Gold/2239. 스도쿠/스도쿠.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();

    static int[][] arr;
    static int SIZE = 9;
    static List<Point> points;

    static class Point {
        int x;
        int y;

        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static boolean dfs(int cnt) {
        if (cnt == points.size()) {
            for (int i = 0; i < SIZE; i++) {
                for (int j = 0; j < SIZE; j++) {
                    sb.append(arr[i][j]);
                }
                sb.append("\n");
            }
            return true;
        }

        // 1 ~ 9 중 어떤 값을 넣을까요?
        for (int i = 1; i < SIZE + 1; i++) {
            int x = points.get(cnt).x;
            int y = points.get(cnt).y;

            // 행, 열, 3x3 행열 모두 조건이 맞다면
            if (checkRow(x, i) && checkCol(y, i) && checkDiag(x, y, i)) {
                arr[x][y] = i;
                if (dfs(cnt + 1)) {
                    return true;
                }

                // 백트래킹
                arr[x][y] = 0;
            }
        }
        return false;
    }

    static boolean checkRow(int x, int i) {
        for (int j = 0; j < SIZE; j++) {
            if (i == arr[x][j]) {
                return false;
            }
        }
        return true;
    }

    static boolean checkCol(int y, int i) {
        for (int j = 0; j < SIZE; j++) {
            if (i == arr[j][y]) {
                return false;
            }
        }
        return true;
    }

    static boolean checkDiag(int x, int y, int i) {
        // 9개의 구역 가장 왼쪽 위 점 찾는 과정
        int nx = x / 3 * 3;
        int ny = y / 3 * 3;

        for (int j = 0; j < 3; j++) {
            for (int k = 0; k < 3; k++) {
                if (i == arr[nx + j][ny + k]) {
                    return false;
                }
            }
        }
        return true;
    }

    public static void main(String[] args) throws Exception {
        arr = new int[SIZE][SIZE];
        points = new ArrayList<>();

        for (int i = 0; i < SIZE; i++) {
            String numbers = br.readLine();
            for (int j = 0; j < SIZE; j++) {
                int num = numbers.charAt(j) - '0';
                arr[i][j] = num;

                if (num == 0) {
                    points.add(new Point(i, j));
                }
            }
        }
        dfs(0);
        System.out.println(sb);
    }
}